WordPress/Classes/Services/MediaCoordinator.swift

```diff
@@ -154,56 +153,88 @@ class MediaCoordinator: NSObject {
-                                        strongSelf.fail(error as NSError, media: media)
-                                    } else {
-                                        // If there was an error and we don't have a media object we just say to the coordinator that one item was finished
-                                        coordinator.finishOneItem()
-                                    }
-                                    return
-                                }
-                                guard let media = media, !media.isDeleted else {
-                                    return
-                                }
-
-                                strongSelf.trackUploadOf(media, analyticsInfo: analyticsInfo)
-
-                                let uploadProgress = strongSelf.uploadMedia(media)
-                                totalProgress.addChild(uploadProgress, withPendingUnitCount: MediaExportProgressUnits.threeQuartersDone)
-        })
-        guard let media = mediaOptional else {
+        let result = service.createMedia(
+            with: asset,
+            blog: post.blog,
+            post: post,
+            thumbnailCallback: { [weak self] media, url in
+                self?.thumbnailReady(url: url, for: media)
+            },
+            completion: { [weak self] media, error in
+                self?.handleMediaImportResult(coordinator: coordinator, totalProgress: totalProgress, analyticsInfo: analyticsInfo, media: media, error: error)
+            }
+        )
+        guard let (media, creationProgress) = result else {
             return nil
         }
+
         processing(media)
-        if let creationProgress = creationProgress {
-            totalProgress.addChild(creationProgress, withPendingUnitCount: MediaExportProgressUnits.quarterDone)
-            coordinator.track(progress: totalProgress, of: media, withIdentifier: media.uploadID)
-        }
+
+        totalProgress.addChild(creationProgress, withPendingUnitCount: MediaExportProgressUnits.quarterDone)
+        coordinator.track(progress: totalProgress, of: media, withIdentifier: media.uploadID)
+
         return media
     }
 
+    /// Create a `Media` instance and upload the asset to the Meida Library.
+    ///
+    /// - SeeAlso: `MediaImportService.createMedia(with:blog:post:receiveUpdate:thumbnailCallback:completion:)`
+    private func addMedia(from asset: ExportableAsset, blog: Blog, post: AbstractPost?, coordinator: MediaProgressCoordinator, analyticsInfo: MediaAnalyticsInfo? = nil) {
+        coordinator.track(numberOfItems: 1)
+        let service = MediaImportService(coreDataStack: coreDataStack)
+        let totalProgress = Progress.discreteProgress(totalUnitCount: MediaExportProgressUnits.done)
+        let creationProgress = service.createMedia(
+            with: asset,
+            blog: blog,
+            post: post,
+            receiveUpdate: { media in
+                self.processing(media)
```

Comment:
Noticed that `self` is not captured weakly here, unlike the other callbacks. I think it's fine in that `self` should always be available when `receiveUpdate` is called, but just checking 😄

Reply:
Addressed in https://github.com/wordpress-mobile/WordPress-iOS/pull/20322/commits/4d4fd8ab941f9f850e961cf679e53b74177372fc — it's good to keep them consistent.